### ISSUE TITLE
Keras BatchNormalization causes "failed to allocate XX bytes from device" error

### ISSUE THREAD

**Issue by harish2704** [ROCmSoftwarePlatform/tensorflow-upstream#361]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
  Linux ubuntu 16.04 ( a1.vega64 instance from https://www.gpueater.com )
- TensorFlow version (use command below):
  1.12.0
- Python version:
  3.5.2
- ROCm/MIOpen version:
```
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.1
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (number of timestamp)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    Intel(R) Core(TM) i3-7100 CPU @ 3.90GHz
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0                                  
  Queue Min Size:          0                                  
  Queue Max Size:          0                                  
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768KB                            
  Chip ID:                 0                                  
  Cacheline Size:          64                                 
  Max Clock Frequency (MHz):3900                               
  BDFID:                   0                                  
  Compute Unit:            4                                  
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    16362668KB                         
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    16362668KB                         
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        TRUE                               
  ISA Info:                
    N/A                      
*******                  
Agent 2                  
*******                  
  Name:                    gfx900                             
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128                                
  Queue Min Size:          4096                               
  Queue Max Size:          131072                             
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      16KB                               
  Chip ID:                 26751                              
  Cacheline Size:          64                                 
  Max Clock Frequency (MHz):1630                               
  BDFID:                   768                                
  Compute Unit:            64                                 
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      FALSE                              
  Wavefront Size:          64                                 
  Workgroup Max Size:      1024                               
  Workgroup Max Size Per Dimension:
    Dim[0]:                  67109888                           
    Dim[1]:                  50332672                           
    Dim[2]:                  0                                  
  Grid Max Size:           4294967295                         
  Waves Per CU:            40                                 
  Max Work-item Per CU:    2560                               
  Grid Max Size per Dimension:
    Dim[0]:                  4294967295                         
    Dim[1]:                  4294967295                         
    Dim[2]:                  4294967295                         
  Max number Of fbarriers Per Workgroup:32                                 
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    8372224KB                          
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        FALSE                              
    Pool 2                   
      Segment:                 GROUP                              
      Size:                    64KB                               
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Alignment:         0KB                                
      Acessible by all:        FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx900          
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Dimension: 
        Dim[0]:                  67109888                           
        Dim[1]:                  1024                               
        Dim[2]:                  16777217                           
      Workgroup Max Size:      1024                               
      Grid Max Dimension:      
        x                        4294967295                         
        y                        4294967295                         
        z                        4294967295                         
      Grid Max Size:           4294967295                         
      FBarrier Max Size:       32                                 
*** Done ***             
```

I was trying to train the this Keras model using ROCm platform. But it failed with following error
```
> /root/pottan-ocr/misc/keras_model.py(15)KerasCrnn()
-> ks = [3, 3, 3, 3, 3, 3, 2]
(Pdb) c
> /root/pottan-ocr/misc/keras_model.py(33)convRelu()
-> cnn.add( BatchNormalization(axis=3, momentum=0.1, epsilon=1e-4, name='batchnorm{0}'.format(i) ))
(Pdb) 
(Pdb) n
2019-03-19 22:01:57.723296: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2019-03-19 22:01:57.723696: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1530] Found device 0 with properties: 
name: Vega 10 XT [Radeon RX Vega 64]
AMDGPU ISA: gfx900
memoryClockRate (GHz) 1.63
pciBusID 0000:03:00.0
Total memory: 7.98GiB
Free memory: 7.73GiB
2019-03-19 22:01:57.723721: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1641] Adding visible gpu devices: 0
2019-03-19 22:01:57.723866: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1051] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-03-19 22:01:57.723877: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1057]      0 
2019-03-19 22:01:57.723883: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1070] 0:   N 
2019-03-19 22:01:57.723913: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1189] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7524 MB memory) -> physical GPU (device: 0, name: Vega 10 XT [Radeon RX Vega 64], pci bus id: 0000:03:00.0)
2019-03-19 22:01:57.769754: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 7.35G (7889485824 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769826: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 6.61G (7100536832 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769851: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 5.95G (6390482944 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769872: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 5.36G (5751434752 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769894: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 4.82G (5176291328 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769915: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 4.34G (4658661888 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769936: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 3.90G (4192795648 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769957: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 3.51G (3773516032 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.769978: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 3.16G (3396164352 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770000: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 2.85G (3056547840 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770025: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 2.56G (2750893056 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770049: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 2.31G (2475803648 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770070: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 2.07G (2228223232 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770092: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1.87G (2005400832 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770113: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1.68G (1804860672 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770134: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1.51G (1624374528 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770155: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1.36G (1461937152 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770177: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1.22G (1315743488 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770201: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1.10G (1184169216 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770224: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 1016.38M (1065752320 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770246: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 914.74M (959177216 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770267: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 823.27M (863259648 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770289: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 740.94M (776933632 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770311: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 666.85M (699240448 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770332: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 600.16M (629316608 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770357: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 540.15M (566385152 bytes) from device: hipError_t(1002)
2019-03-19 22:01:57.770380: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 486.13M (509746688 bytes) from device: hipError_t(1002)
terminate called after throwing an instance of 'std::runtime_error'
  what():  No device code available for function: _ZN10tensorflow7functor28FillPhiloxRandomKernelLaunchINS_6random19UniformDistributionINS2_12PhiloxRandomEfEEEEvS4_PNT_17ResultElementTypeExS6_
Aborted (core dumped)
```

**Code to reproduce the issue**
```python
python3 misc/run_keras.py   --traindata ../datasets/validate.txt.gz  --valdata ../datasets/validate.txt.gz   --batchSize 64 --traindata_limit $(( 1999*64 )) --valdata_limit 128 --nh 64
```
*Complete shell script to automate the process:* https://github.com/harish2704/harish2704.github.io/blob/master/for-pottan-ocr/boot_gpueater.sh

* The line which caused this bug* : https://github.com/harish2704/pottan-ocr/blob/108234c589533608b9a7ced7f95906cd1c6cf19d/misc/keras_model.py#L31

**Comment by sunway513** [comment]
@harish2704 The error message looks like related to your local configuration. 
Could you try to run the HIP unit tests? Link is here:
https://github.com/ROCm-Developer-Tools/HIP/tree/master/tests

**Comment by sunway513** [comment]
Besides, could you also review the following comment from @jlgreathouse ?
https://github.com/RadeonOpenCompute/ROCm/issues/622#issuecomment-441941594
The failure signature seems related to your case.

**Comment by harish2704** [comment]
Hi @sunway513 : You are right. It seems like a configuration problem of the system I am using.
The issue disappeared when I tried to replicate the same on a different machine.

I was not expecting an hardware configuration issue from a service provider who officially provides ROCm enabled GPU instances . That is why I didn't checked for the possibility of a configuration issue in the first time.

Thanks!!
